pkg/services/pluginsintegration/clientmiddleware/tracing_header_middleware.go

```diff
@@ -0,0 +1,94 @@
+package clientmiddleware
+
+import (
+	"context"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/services/contexthandler"
+	"github.com/grafana/grafana/pkg/services/query"
+)
+
+// NewTracingHeaderMiddleware creates a new plugins.ClientMiddleware that will
+// populate useful tracing headers on outgoing plugins.Client and HTTP
+// requests.
+// Tracing headers are X-Datasource-Uid, X-Dashboard-Uid,
+// X-Panel-Id, X-Grafana-Org-Id.
+func NewTracingHeaderMiddleware() plugins.ClientMiddleware {
+	return plugins.ClientMiddlewareFunc(func(next plugins.Client) plugins.Client {
+		return &TracingHeaderMiddleware{
+			next: next,
+		}
+	})
+}
+
+type TracingHeaderMiddleware struct {
+	next plugins.Client
+}
+
+func (m *TracingHeaderMiddleware) applyHeaders(ctx context.Context, pCtx backend.PluginContext, req interface{}) context.Context {
+	reqCtx := contexthandler.FromContext(ctx)
+	// If no HTTP request context then skip middleware.
+	if req == nil || reqCtx == nil || reqCtx.Req == nil {
+		return ctx
+	}
+
+	var headersList = []string{query.HeaderPanelID, query.HeaderDashboardUID, query.HeaderDatasourceUID, `X-Grafana-Org-Id`}
+
+	switch t := req.(type) {
+	case *backend.QueryDataRequest:
+		for _, httpHeaderName := range headersList {
+			t.Headers[httpHeaderName] = reqCtx.Req.Header.Get(httpHeaderName)
```

Comment:
Shouldn't populate the header if it's not set in incoming request. Given my explanation of the PR I'm working on I think SetHTTPHeader should be used since this is new feature for backend plugins and `backend.QueryDataRequest.Headers` is not entirely for HTTP headers - it's for environment/metadata about the reuqest.

Same applies for CheckHealthRequest below.